etna/ciim/utils.py

```diff
@@ -14,6 +15,31 @@ def underscore_to_camelcase(word, lower_first=True):
     return result
 
 
+def extract(source: Any, key: str, default: Any = "", fail_loudly: bool = False) -> Any:
+    """
+    Utilises Django template engine's variable resolution logic to extract
+    values from unpredictable nested data structures using a simple '.'
+    separated string to indicate keys, attributes, methods and indexes to be
+    traversed or called.
+    """
+    variable_str = f"source.{key}"
+    try:
+        # attempt regular initialization to trigger validation
+        variable_node = Variable(variable_str)
+    except TemplateSyntaxError:
+        # the template engine doesn't like segments that start with '_', but
+        # that is not an issue here, so let's bypass that validation
+        variable_node = Variable("source")
+        variable_node.var = variable_str
+        variable_node.lookups = tuple(variable_str.split("."))
+    try:
+        return variable_node.resolve({"source": source})
+    except VariableDoesNotExist:
+        if fail_loudly:
+            raise
+        return default
```

Comment:
Can we have tests around VariableDoesNotExist:=, fail_loudly=True and default?

Reply:
Yes

Reply:
This is irrelevant now, as `VariableDoesNotExist` is not used